src/storage/leveldb/address_index.ts

```diff
@@ -0,0 +1,184 @@
+
+  async close(): Promise<void> {
+    await this.addressesDB.db.close();
+  }
+
+  async checkVersion(): Promise<void> {
+    const db = this.addressesDB.db;
+    try {
+      const dbVersion = await db.get('version');
+      if (this.indexVersion !== dbVersion) {
+        throw new Error(`Database version mismatch for ${this.constructor.name}: database version (${dbVersion}) expected version (${this.indexVersion})`);
+      }
+    } catch (err: unknown) {
+      if (errorCodeOrNull(err) === KEY_NOT_FOUND_CODE) {
+        // This is a new db, add version and return
+        await db.put('version', this.indexVersion);
+        return;
+      }
+
+      throw err;
+    }
+  }
+
+  async validate(): Promise<AddressIndexValidateResponse> {
+    // Clear metadata since we cannot guarantee the validity
+    await this.addressesMetaDB.clear();
+
+    const ret: AddressIndexValidateResponse = {firstIndex: Infinity, lastIndex: -1};
+    // Iterate on all addresses and check that we have a corresponding index entry
+    for await (const [key, value] of this.addressesDB.iterator()) {
+      if (key !== value.base58) {
+        throw new Error('Inconsistent database');
+      }
+      if (value.bip32AddressIndex < ret.firstIndex) {
+        ret.firstIndex = value.bip32AddressIndex;
+      }
+      if (value.bip32AddressIndex > ret.lastIndex) {
+        ret.lastIndex = value.bip32AddressIndex;
+      }
+
+      // check that we have an index
+      try {
+        const addressFromIndex = await this.addressesIndexDB.get(_index_key(value.bip32AddressIndex));
+        if (value.base58 !== addressFromIndex) {
+          throw new Error('Inconsistent database');
+        }
+      } catch(err: unknown) {
+        if (errorCodeOrNull(err) === KEY_NOT_FOUND_CODE) {
+          // Create if index is missing
+          await this.addressesIndexDB.put(_index_key(value.bip32AddressIndex), value.base58);
+          continue;
+        }
+        throw err;
+      }
+    }
+    return ret;
+  }
+
+  async getAddressInfo(base58: string): Promise<IAddressInfo | null> {
+    try {
+      return await this.addressesDB.get(base58);
+    } catch(err: unknown) {
+      if (errorCodeOrNull(err) === KEY_NOT_FOUND_CODE) {
+        return null;
+      }
+      throw err;
+    }
+  }
+
+  async addressExists(base58: string): Promise<boolean> {
+    return (await this.getAddressInfo(base58)) !== null;
+  }
+
+  async * addressIter(): AsyncGenerator<IAddressInfo> {
+    for await (const address of this.addressesIndexDB.values()) {
+      const info = await this.getAddressInfo(address);
+      if (info === null) {
+        throw new Error('Inconsistent database');
+      }
+      yield info;
+    }
+  }
+
+  async setAddressMeta(address: string, meta: IAddressMetadata): Promise<void> {
+    const dbMeta: IAddressMetadataAsRecord = {numTransactions: meta.numTransactions, balance: {}};
+    for (const [uid, balance] of meta.balance.entries()) {
+      dbMeta.balance[uid] = balance;
+    }
```

Comment:
`IAddressMetadata` is defined as:

```ts
export interface IAddressMetadata {
  numTransactions: number;
  balance: Map<string, IBalance>;
}
```
Unfortunately, `Map` is not supported by the json encoding lib of leveldb, so I was required to create a  `IAddressMetadataAsRecord` which uses `Record<string, IBalance>` instead of `Map<string, IBalance>` and I have to change between these types manually when required.